ingestion/src/metadata/profiler/profiler/interface/profiler_protocol.py

```diff
@@ -0,0 +1,239 @@
+#  Copyright 2021 Collate
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  http://www.apache.org/licenses/LICENSE-2.0
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""
+Interfaces with database for all database engine
+supporting sqlalchemy abstraction layer
+"""
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Union
+
+from pydantic import BaseModel
+from sqlalchemy import Column, MetaData
+from typing_extensions import Self
+
+from metadata.generated.schema.entity.data.table import PartitionProfilerConfig, Table
+from metadata.generated.schema.entity.services.connections.database.datalakeConnection import (
+    DatalakeConnection,
+)
+from metadata.generated.schema.entity.services.databaseService import DatabaseConnection
+from metadata.generated.schema.metadataIngestion.databaseServiceProfilerPipeline import (
+    DatabaseServiceProfilerPipeline,
+)
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.profiler.api.models import ProfileSampleConfig, TableConfig
+from metadata.profiler.metrics.registry import Metrics
+from metadata.utils.partition import get_partition_details
+
+
+class ProfilerInterfaceArgs(BaseModel):
+    """Profiler Interface Args Model"""
+
+    service_connection_config: Any
+    sqa_metadata_obj: Optional[MetaData]
+    ometa_client: Optional[OpenMetadata]
+    thread_count: Optional[float]
+    table_entity: Optional[Union[Table, Any]]
+    profile_sample_config: Optional[ProfileSampleConfig] = None
+    table_sample_query: Optional[Union[int, str]]
+    table_partition_config: Optional[PartitionProfilerConfig]
+    timeout_seconds: Optional[int]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class ProfilerProtocol(ABC):
+    """Protocol interface for the profiler processor"""
+
+    _profiler_type: Optional[str] = None
+    subclasses = {}
+
+    def __init_subclass__(cls, *args, **kwargs) -> None:
+        """Hook to map subclass objects to profiler type"""
+        super().__init_subclass__(*args, **kwargs)
```

Comment:
I might need an example here to follow this 🤔 Maybe worth it to have it as part of the docstring itself

Reply:
Basically when a class inherit from another class `__init_subclass__` is called on the parent. Our interfaces inherit from the protocol. These subclass will be registered in the subclass dict with the key as the connection type and the value as the interface class itself.

When we call `.create()` we will instantiate the class based on the connection type of the service

```python
        return cls.subclasses[_profiler_type](
            service_connection_config=service_connection_config,
            ometa_client=ometa_client,
            thread_count=thread_count,
            entity=entity,
            profile_sample_config=profile_sample_config,
            sample_query=sample_query,
            table_partition_config=table_partition_config,
            timeout_seconds=timeout_seconds,
            **kwargs,
        )
```

If we want to add a dashboard service profiler, we should only have to create a new interface class that inherits from the protocol.